##1.1 Installing librat
###Phil Wilkes (UCL/NCEO) <a href=mailto:p.wilkes@ucl.ac.uk>p.wilkes@ucl.ac.uk</a>

This tutorial runs through the basics of running librat.  Material is taken from <a href=https://librat.wikispaces.com/>librat wikispace</a>
<p>Librat is usually run from the command line, but to make things clearer, easier to run and more interactive, this is Jupyter work book uses Python to call librat using the subprocess function.

##Prerequisites
Imagemajick which is available <a href

Installing librat can be done on unix-like operating system (tested on Solaris, various linux, Mac OS X, etc.). <b>Librat will not install on a Windows OS</b>. To install librat download the <a href=http://www2.geog.ucl.ac.uk/~plewis/librat/configure>configure script</a>.  Then change the permission to make it an executable, this is done on the command line,
<code><span style="background-color:lightgrey">
chmod +x configure
</span></code>

Install librat with;
<code><span style="background-color:lightgrey">
./configure
</span></code>

There are also a few environment variables that need to be updated:
<code><span style="background-color:lightgrey">
export BPMS=${HOME}/bpms
export PATH=$BPMS/bin/x86_64:$BPMS/bin/csh:$PATH
</span></code>

For Mac users:
<code><span style="background-color:lightgrey">
export DYLD_LIBRARY_PATH=${BPMS}/src/lib
</span></code>

For Linux ueser:
<code><span style="background-color:lightgrey">
export LD_LIBRARY_PATH=${BPMS}/src/lib
</span></code>

These can be added to your <code><span style="background-color:lightgrey">.bash_profile</span></code> or <code><span style="background-color:lightgrey">.bashrc</span></code>.  For this iPython session now we can run this below.

In [3]:
import platform
import os

os.environ['BPMS'] = os.path.join(os.path.expanduser("~"), 'bpms')
os.environ['PATH'] += os.path.join(os.environ['BPMS'], 'bin/x86_64')
os.environ['PATH'] += os.path.join(os.environ['BPMS'], 'bin/csh')

if platform.system() == 'Darwin': # A Mac:
    os.environ['DYLD_LIBRARY_PATH'] = os.path.join(os.environ['BPMS'], 'src/lib')
else:
    os.environ['LD_LIBRARY_PATH'] = os.path.join(os.environ['BPMS'], 'src/lib')

We are now ready to test if librat is working

In [10]:
# this is an inbulit simple test we can run...
import subprocess
os.chdir(os.path.join(os.environ['BPMS'], 'src/start/src/lib/src/start'))
stdout = subprocess.Popen('./testMe', stdout=subprocess.PIPE)
for line in stdout.stdout:
    print line.strip()

BPMS environment variable set to /Users/phil/bpms
this result
RTD 0
order: 0	intersection point:	0.000000 0.000000 1.000010
ray length:		98.999990
intersection material:	6
sun 0:			1 reflectance
lengthToSun: 9.000000 angleToSun 0.000000
direct:			0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
sky  :			reflectance
diffuse:		0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
wavebands: 0.000000 1.000000 2.000000 3.000000 4.000000 5.000000 6.000000
x: -0.499900 0.500100
y: -19.996000 20.004000
z: -0.999800 1.000200
bbox centre @ 0.000100 0.004000 0.000200
diff of file:


If the output above looks like this:

<code>
this result
<code>===========</code>
RTD 0
order: 0	intersection point:	0.000000 0.000000 1.000010
ray length:		98.999990
intersection material:	6
sun 0:			1 reflectance
lengthToSun: 9.000000 angleToSun 0.000000
direct:			0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
sky  :			reflectance
diffuse:		0.000000e+00 0.000000e+00 0.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00
wavebands: 0.000000 1.000000 2.000000 3.000000 4.000000 5.000000 6.000000
x: -0.499900 0.500100
y: -19.996000 20.004000
z: -0.999800 1.000200
bbox centre @ 0.000100 0.004000 0.000200
diff of file:
=============</code>

Then librat has been installed correctly

Librat can be run from the command line - here are some <a href=http://librat.wikispaces.com/Basic+examples>basic examples</a> - luckily though <a href=mailto:https://www.researchgate.net/profile/John_Armston2>John Armston</a> has created some useful Python functions that call librat. These can be downloaded from <a href=http://librat.wikispaces.com/Tools>here</a>.

To call these functions after downloading, you will need to modify your environment variables by adding the location of the tools to the PYTHONPATH variable. In Linux this is done by

<code>export PYTHONPATH=/path/to/librat/tools:$PYTHONPATH</code>

You can then...

In [4]:
from librattools import lidarsurvey


In [6]:
lidarsurvey.terrestrialConfig()

In [5]:
os.environ['PYTHONPATH']

'/Users/phil/FLIGHT75'

In [8]:
sys.path.append('/Users/phil/Dropbox/UCL/librat/librattools-0.1')